core-rust/state-manager/src/store/rocks_db.rs

```diff
@@ -78,217 +79,286 @@ use radix_engine::types::{
         if let LedgerTransaction::User(notarized_transaction) = &transaction {
-            let key = get_transaction_intent_key(&notarized_transaction.intent_hash());
-            let existing_intent_option = self
+            let existing_intent_hash_mapping_opt = self
                 .db_txn
-                .get_for_update(key, true)
+                .get_for_update_cf(
+                    self.cf_handle(&StateVersionByTxnIntentHash),
+                    notarized_transaction.intent_hash(),
+                    true,
+                )
                 .expect("RocksDB: failure to read intent hash");
-            if let Some(existing_intent_hash) = existing_intent_option {
+            if let Some(existing_intent_hash_mapping) = existing_intent_hash_mapping_opt {
                 panic!(
-                    "Attempted to save intent hash which already exists: {:?}",
-                    IntentHash::from_raw_bytes(existing_intent_hash.try_into().unwrap())
+                    "Attempted to save intent hash {:?} which already exists at state version {:?}",
+                    notarized_transaction.intent_hash(),
+                    u64::from_be_bytes(existing_intent_hash_mapping.try_into().unwrap())
                 );
             }
             self.db_txn
-                .put(
-                    get_transaction_intent_key(&notarized_transaction.intent_hash()),
-                    transaction.get_hash().as_ref(),
+                .put_cf(
+                    self.cf_handle(&StateVersionByTxnIntentHash),
+                    notarized_transaction.intent_hash(),
+                    state_version.to_be_bytes(),
                 )
                 .expect("RocksDB: failure to put intent hash");
+
+            self.db_txn
+                .put_cf(
+                    self.cf_handle(&StateVersionByTxnUserPayloadHash),
+                    notarized_transaction.user_payload_hash(),
+                    state_version.to_be_bytes(),
+                )
+                .expect("RocksDB: failure to put user payload hash");
         }
-        let transaction_key = get_transaction_key(&transaction.get_hash());
-        // TODO: We really need to re-work how we store this data!
-        // For now, panic if we have duplicate transactions to prevent incorrect receipt/identifiers being saved
-        // There shouldn't be any possibility of this happening at the moment, so better to panic if it does as
-        // it indicates a bug.
-        let transaction_already_exists = self
-            .db_txn
-            .get(transaction_key)
-            .expect("RocksDB: failure to check for transaction presence")
-            .is_some();
-        if transaction_already_exists {
-            panic!("Attempt to persist a duplicate transaction payload - which isn't supported at the moment in our database structure: {:?}", transaction)
-        }
+
```

Comment:
Not sure if we still want to check for ledger payload hash duplicates? I haven't added it yet in this PR as it seems unnecessary (user txns are covered by the intent hash check and there should be no more duplicates in validator/system txns [unless there's a bug, but I'm not convinced that keeping an extra query here is worth it])

Reply:
> Why do we actually need an index on user_payload_hash or ledger_payload_hash?

Hah, good question. I guess they're both not really needed, but might be useful to have them around for e.g. debugging?

> Is "no duplicates for ledger payload hash" an invariant we actually need?

Another good question :thinking: intuition tells me "yes", but I can't really come up with an example to justify it. I don't think we rely on this invariant anywhere in the node

Reply:
If it's easy to re-run the benchmark, I'd be curious: if we remove maintaining these indices (ie remove insertions on create for now), do we see a noticeable commit speedup?

Reply:
There's almost no difference

Reply:
I guess it makes sense then to keep the index and add an extra ledger payload duplicate check to ensure correctness